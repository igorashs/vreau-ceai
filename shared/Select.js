import styled from 'styled-components';
import ArrowSvg from 'assets/icons/arrow.svg';
import { Label } from '@/shared/Label';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: 7px;
`;

const SelectWrapper = styled.div`
  position: relative;
  align-self: flex-start;

  ${({ fullWidth }) => fullWidth && 'width: 100%;'}
`;

const StyledSelect = styled.select`
  width: 100%;
  padding: 7px 38px 7px 14px;
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  text-decoration: none;
  appearance: none;
  cursor: pointer;

  :hover {
    background-color: var(--accent);
  }

  option {
    padding: 7px 14px;
  }
`;

const StyledSvg = styled(ArrowSvg)`
  position: absolute;
  transform: rotate(90deg);
  color: var(--text-light);
  right: 7px;
  top: calc(50% - 12px);
  pointer-events: none;
`;

export function Select({
  name,
  label,
  error,
  passRef,
  id = name,
  fullWidth,
  disabled,
  children
}) {
  return (
    <Wrapper>
      <Label htmlFor={id} error={error}>
        {error || label}
      </Label>
      <SelectWrapper fullWidth={fullWidth}>
        <StyledSelect name={name} ref={passRef} id={id} disabled={disabled}>
          {children}
        </StyledSelect>
        <StyledSvg />
      </SelectWrapper>
    </Wrapper>
  );
}
