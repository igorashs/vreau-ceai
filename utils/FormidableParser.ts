import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
