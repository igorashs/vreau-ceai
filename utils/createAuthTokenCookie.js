import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
