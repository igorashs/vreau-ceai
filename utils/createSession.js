import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
