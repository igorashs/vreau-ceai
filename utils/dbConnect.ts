import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
