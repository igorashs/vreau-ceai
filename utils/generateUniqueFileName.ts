import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
