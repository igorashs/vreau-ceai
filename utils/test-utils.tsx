import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
