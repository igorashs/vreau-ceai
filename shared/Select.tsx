import styled from 'styled-components';
import ArrowSvg from 'assets/icons/arrow.svg';
import { Label } from '@/shared/Label';

interface SelectWrapperProps {
  fullWidth?: boolean;
}

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: 7px;
`;

const SelectWrapper = styled.div<SelectWrapperProps>`
  position: relative;
  align-self: flex-start;

  ${({ fullWidth }) => fullWidth && 'width: 100%;'}
`;

const StyledSelect = styled.select`
  width: 100%;
  padding: 7px 38px 7px 14px;
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  text-decoration: none;
  appearance: none;
  cursor: pointer;

  :hover {
    background-color: var(--accent);
  }

  option {
    padding: 7px 14px;
  }
`;

const StyledSvg = styled(ArrowSvg)`
  position: absolute;
  transform: rotate(90deg);
  color: var(--text-light);
  right: 7px;
  top: calc(50% - 12px);
  pointer-events: none;
`;

interface SelectProps {
  name: string;
  label: string;
  error?: boolean;
  id?: string;
  fullWidth?: boolean;
  disabled?: boolean;
  passRef:
    | ((instance: HTMLSelectElement) => void)
    | React.RefObject<HTMLSelectElement>;
}

export const Select = ({
  name,
  label,
  error = false,
  passRef = null,
  id = name,
  fullWidth = false,
  disabled = false,
  children,
}: React.PropsWithChildren<SelectProps>) => {
  return (
    <Wrapper>
      <Label htmlFor={id} error={error}>
        {error || label}
      </Label>
      <SelectWrapper fullWidth={fullWidth}>
        <StyledSelect name={name} ref={passRef} id={id} disabled={disabled}>
          {children}
        </StyledSelect>
        <StyledSvg />
      </SelectWrapper>
    </Wrapper>
  );
};
