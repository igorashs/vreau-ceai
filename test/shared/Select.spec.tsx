import { render } from '@/utils/test-utils';
import Select from '@/shared/Select';

describe('Select', () => {
  it('renders correctly', () => {
    const name = 'select';
    const label = 'select label';

    const { container, getByRole, getByText } = render(
      <Select name={name} label={label}>
        <option value="op1">op1</option>
        <option value="op2">op2</option>
      </Select>,
    );

    const select = getByRole('combobox');

    expect(select).toHaveAttribute('name', name);
    expect(select).toHaveAttribute('id', name);
    expect(getByText(label)).toBeVisible();
    expect(getByText('op1')).toBeVisible();
    expect(getByText('op2')).toBeVisible();
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('fullWidth prop', () => {
    it('renders with fullWidth', () => {
      const name = 'select';
      const label = 'select label';

      const { getByRole } = render(
        <Select name={name} label={label} fullWidth>
          <option value="op1">op1</option>
          <option value="op2">op2</option>
        </Select>,
      );

      expect(getByRole('combobox').parentElement).toHaveStyleRule(
        'width',
        '100%',
      );
    });
  });
});
