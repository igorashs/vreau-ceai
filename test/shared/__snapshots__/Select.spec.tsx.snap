// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders correctly 1`] = `
.c1 {
  font-weight: 500;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 7px;
}

.c2 {
  position: relative;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.c3 {
  width: 100%;
  padding: 7px 38px 7px 14px;
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  cursor: pointer;
}

.c3:hover {
  background-color: var(--accent);
}

.c3 option {
  padding: 7px 14px;
}

.c4 {
  position: absolute;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
  color: var(--text-light);
  right: 7px;
  top: calc(50% - 12px);
  pointer-events: none;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="select"
  >
    select label
  </label>
  <div
    class="c2"
  >
    <select
      class="c3"
      id="select"
      name="select"
    >
      <option
        value="op1"
      >
        op1
      </option>
      <option
        value="op2"
      >
        op2
      </option>
    </select>
    <svg
      class="c4"
    />
  </div>
</div>
`;
